core/types/receipt.go

```diff
@@ -87,29 +100,35 @@ func (r *Receipt) DecodeRLP(s *rlp.Stream) error {
 	if err != nil {
 		return err
 	}
-	list, _, err := rlp.SplitList(raw)
+	content, _, err := rlp.SplitList(raw)
 	if err != nil {
 		return err
 	}
-	items, err := rlp.CountValues(list)
+	kind, cnt, _, err := rlp.Split(content)
 	if err != nil {
 		return err
 	}
-	// Deserialize based on the number of content items
-	switch items {
-	case 3:
-		// Metropolis receipts have 3 components
+	// Deserialize based on the first component type.
+	switch {
+	case kind == rlp.Byte || kind == rlp.String && len(cnt) == 0:
+		// The first component of metropolis receipts is Byte
+		// or empty String(byte with 0x00 value).
 		var metro metropolisReceiptRLP
 		if err := rlp.DecodeBytes(raw, &metro); err != nil {
 			return err
 		}
+		if metro.Status == receiptStatusSuccessful {
+			r.Failed = false
+		} else {
+			r.Failed = true
+		}
```

Comment:
You can set the failed flag with a single statement: `r.Failed = metro.Status == receiptStatusFailed`